### PR TITLE
Compute derivative of basis function via chain rule.

### PR DESCRIPTION
**Pull request by Aerojspark**
This commit modifies the way to compute the derivative of basis function via chain rule.

### CODE REVIEW

File: pyfr/mputil.py [hunk 1]
```diff
@@ -25,6 +25,16 @@ def jacobi(n, a, b, z):
     return j
 
 
+def derivative(f, x):
```
Comment (FreddieWitherden) [inline pyfr/mputil.py]: Why do we have this function?  mpmath already provides numerical differentiation routines and, moreover, why do we not have a `jacobi_diff` method which computes the first differential of the Jacobi polynomials analytically?

File: pyfr/polys.py [hunk 2]
```diff
@@ -167,6 +252,41 @@ def ortho_basis_at_mp(self, p, q, r):
...
+
+                qij = -i*tmp*fi*gj
+                qij += (1 + a)*tmp*dfi*gj
+                qij += (1 - b)**i*fi*dgj
+
+                rij = (1 - b)**i*fi*gj
+
+                pab.append([cij*pij, cij*qij, cij*rij])
+
+        h = lambda x: jacobi(self.order - 1, 0, 0, x)
```
Comment (FreddieWitherden) [inline pyfr/polys.py]: This is the same as `f`.

File: pyfr/polys.py [hunk 3]
```diff
@@ -167,6 +252,41 @@ def ortho_basis_at_mp(self, p, q, r):
...
+    def jac_ortho_basis_at_mp(self, p, q, r):
+        a = 2*(1 + p)/(1 - q) - 1 if q != 1 else -1
+        b = q
+        c = r
+
+        f = lambda x: jacobi(self.order - 1, 0, 0, x)
+
+        pab = []
+        for i, (fi, dfi) in enumerate(zip(f(a), derivative(f, a))):
+            g = lambda x: jacobi(self.order - i - 1, 2*i + 1, 0, x)
```
Comment (FreddieWitherden) [inline pyfr/polys.py]: Why is the being redefined each iteration?

File: pyfr/mputil.py (hunk 1 above)
Comment (Aerojspark) [inline pyfr/mputil.py]: I think mp.diff cannot work for array of function. It is easier to implement than jacobi_diff.

File: pyfr/polys.py (hunk 3 above)
Comment (Aerojspark) [inline pyfr/polys.py]: I think that the function g depends on index i, so I redefine for this loop.

Comment (FreddieWitherden) [inline pyfr/polys.py]: Variables are resolved at call time.

```
In [2]: f = lambda x: x + i

In [3]: i = 1

In [4]: f(1)
Out[4]: 2

In [5]: i = 3

In [6]: f(2)
Out[6]: 5
```

File: pyfr/mputil.py [hunk 4]
```diff
@@ -25,6 +25,16 @@ def jacobi(n, a, b, z):
     return j
 
 
+def jacobi_diff(n, a, b, z):
+    dj = [0]
+
+    if n >= 1:
+        dj += [jp*mp.mpf((i + a + b + 2)/2)
```
Comment (FreddieWitherden) [inline pyfr/mputil.py]: `.extend(x for x in y)`

File: pyfr/polys.py [hunk 5]
```diff
@@ -204,6 +316,44 @@ def ortho_basis_at_mp(self, p, q, r):
...
+
+                    tmp = (1 - c)**(i + j-1) if i + j > 0 else 1
+
+                    pijk = 2*tmp*dfi*gj*hk
+
+                    qijk = 2*tmp*fi*dgj*hk
+
+                    rijk = a*tmp*dfi*gj*hk
+                    rijk += b*tmp*fi*dgj*hk
+                    rijk += -(i+j)*tmp*fi*gj*hk + (1 - c)**(i+j)*fi*gj*dhk
```
Comment (FreddieWitherden) [inline pyfr/polys.py]: Merge into a single statement.

File: pyfr/polys.py [hunk 6]
```diff
@@ -167,6 +246,41 @@ def ortho_basis_at_mp(self, p, q, r):
...
+            for j, (gj, dgj) in enumerate(zip(g, dg)):
+                cij = mp.sqrt((2*i + 1)*(2*i + 2*j + 2)) / 2**(i + 1)
+
+                tmp = (1 - b)**(i - 1) if i > 0 else 1
+
+                pij = 2*tmp*gj*dfi
+
+                qij = -i*tmp*fi*gj
+                qij += (1 + a)*tmp*dfi*gj
+                qij += (1 - b)**i*fi*dgj
```
Comment (FreddieWitherden) [inline pyfr/polys.py]: Merge into a single statement.

File: pyfr/polys.py [hunk 7]
```diff
@@ -135,6 +174,48 @@ def ortho_basis_at_mp(self, p, q, r):
...
+                    pijk = 4*tmp2*gj*hk*dfi
+
+                    qijk = -2*i*tmp2*fi*gj*hk
+                    qijk += 2*(1 + a)*tmp2*gj*hk*dfi
+                    qijk += 2*tmp1*(1 -b)**i*fi*hk*dgj
+
+                    rijk = -(i*(1 + b)*tmp2 + (i + j)*tmp1*(1 - b)**i)*fi*gj*hk
+                    rijk += 2*(1 + a)*tmp2*gj*hk*dfi
+                    rijk += (1 + b)*tmp1*(1 - b)**i*fi*hk*dgj
+                    rijk += (1 - c)**(i + j)*(1 - b)**i*fi*gj*dhk
```
Comment (FreddieWitherden) [inline pyfr/polys.py]: Merge into a single statement.

File: pyfr/polys.py [hunk 8]
```diff
@@ -135,6 +174,48 @@ def ortho_basis_at_mp(self, p, q, r):
...
+                    cijk = cij*ck
+
+                    tmp1 = (1 - c)**(i + j - 1) if i + j > 0 else 1
+                    tmp2 = tmp1*(1 - b)**(i - 1) if i > 0 else 1
+
+                    pijk = 4*tmp2*gj*hk*dfi
+
+                    qijk = -2*i*tmp2*fi*gj*hk
+                    qijk += 2*(1 + a)*tmp2*gj*hk*dfi
+                    qijk += 2*tmp1*(1 -b)**i*fi*hk*dgj
```
Comment (FreddieWitherden) [inline pyfr/polys.py]: Merge into a single statement.

File: pyfr/polys.py [hunk 9]
```diff
@@ -86,6 +88,33 @@ def ortho_basis_at_mp(self, p, q):
...
+            for j, (gj, dgj) in enumerate(zip(g, dg)):
+                cij = mp.sqrt((2*i + 1)*(2*i + 2*j + 2)) / 2**(i + 1)
+
+                tmp = (1 - b)**(i - 1) if i > 0 else 1
+
+                pij = 2*tmp*gj*dfi
+
+                qij = -i*tmp*fi*gj
+                qij += (1 + a)*tmp*dfi*gj
+                qij += (1 - b)**i*fi*dgj
```
Comment (FreddieWitherden) [inline pyfr/polys.py]: Merge into a single statement.

File: pyfr/mputil.py [hunk 10]
```diff
@@ -25,6 +25,16 @@ def jacobi(n, a, b, z):
...
+    dj = [0]
+
+    if n >= 1:
+        dj.extend(jp*mp.mpf((i + a + b + 2)/2)
+                  for i, jp in enumerate(jacobi(n - 1, a + 1, b + 1, z)))
+
+    return dj
+
+
 def jacobian(f, x):
```
Comment (FreddieWitherden) [inline pyfr/mputil.py]: We should look to ditch this function.

File: pyfr/polys.py [hunk 11]
```diff
@@ -204,6 +309,41 @@ def ortho_basis_at_mp(self, p, q, r):
...
+                )
+
+                for k, (hk, dhk) in enumerate(zip(h, dh)):
+                    ck = mp.sqrt(2*(k + j + i) + 3)
+
+                    tmp = (1 - c)**(i + j-1) if i + j > 0 else 1
+
+                    pijk = 2*tmp*dfi*gj*hk
+                    qijk = 2*tmp*fi*dgj*hk
+                    rijk = (tmp*(a*dfi*gj + b*fi*dgj - (i+j)*fi*gj)*hk
```
Comment (FreddieWitherden) [inline pyfr/polys.py]: Spacing around ` + `.

File: pyfr/polys.py [hunk 12]
```diff
@@ -167,6 +243,37 @@ def ortho_basis_at_mp(self, p, q, r):
...
+                rij = (1 - b)**i*fi*gj
+
+                pab.append([cij*pij, cij*qij, cij*rij])
+
+        sk = [mp.sqrt(k + 0.5) for k in range(self.order)]
+        hc = [s*jp for s, jp in zip(sk, jacobi(self.order - 1, 0, 0, c))]
+        dhc = [s*jp for s, jp in zip(sk, jacobi_diff(self.order - 1, 0, 0, c))]
+
+        return [[pij*hk, qij*hk, rij*dhk]
+                for (pij, qij, rij) in pab for (hk, dhk) in zip(hc, dhc)]
```
Comment (FreddieWitherden) [inline pyfr/polys.py]: Parenthesis.

**Comment by FreddieWitherden** [comment]
Assuming all of your tests pass I am happy to merge.

**Pull request merged**